### PR TITLE
Fix loc for repeating text in text_splitter

### PR DESCRIPTION
**Pull request by tonisives**
This PR fixes an issue where the Document `loc` is wrong if there are repeating blocks. This happens especially obviously in code files where there are similar closing blocks. For example `\n    }`. 

The current code erroneously sets the `loc` to more lines there are lines in the file. This is because it doesn't use the lastIndex of the chunk, but instead the first occurrence.

Might also fix issue #1357

### CODE REVIEW

**Comment by jacoblee93** [comment]
Makes sense but breaks a few other test cases :( Can you fix?

**Comment by tonisives** [comment]
This solution needs some more work. My LOC is still not precise in some cases.

Probably yeah, fixing the tests would be a good start

**Comment by yyvess** [comment]
@tonisives 
I created a similarly PR (sorry I see your after) bit more simple to fix the issue about new lines on overlap and on beginning of the content 
https://github.com/langchain-ai/langchainjs/pull/2779

About issues with multiple chunk with same content that I understand you try to fix by "find the correct loc by brute force".

I think seen better to completely rewrite the split function to compute lines number during the split and not try to find line number after the splitting as it actually.

But that require a big refactoring ....

**Comment by tonisives** [comment]
I added a brute force loc fixer after splitting documents. [add fix loc method](https://github.com/langchain-ai/langchainjs/pull/1977/commits/850a30f5846cd639c9cc2d8e1f710d7fceffe8c0)

It is obviously inefficient and the loc should be correct from the initial parsing. I think the problem comes from `splitText`, which doesn’t return \n characters, which are used to count the loc. It’s a confusing function and I don’t have time to debug it.

-- 

Generally a rewrite sounds like a good idea. Starting with proper test cases for all of the features and then separate functions for each of the features (chunks, overlaps, mapping, etc). Currently the code is hard to follow and I think not all features are properly tested.

But it's a lot of work and I might just write a personal one for my use case only.

**Comment by tonisives** [comment]
verified fixed in https://github.com/langchain-ai/langchainjs/pull/2779  

https://github.com/langchain-ai/langchainjs/releases/tag/0.0.161
